# NFHS-5 Data Cleaning

This notebook documents the data cleaning and preprocessing steps performed on the National Family Health Survey (NFHS-5) district-level dataset.

The NFHS-5 survey was conducted across India during **2019–2021** under the stewardship of the International Institute for Population Sciences (IIPS), Mumbai, with technical support from the ICF. The survey provides comprehensive information on population, health, nutrition, and socioeconomic indicators at the national, state, and district levels.

The objective of this notebook is to prepare the dataset for exploratory data analysis by identifying and addressing data quality issues. The cleaning workflow includes:

* Initial dataset inspection and validation
* Column name standardization
* Duplicate record detection
* Missing value analysis
* Handling variables with excessive missing values
* Missing value imputation
* Final dataset validation and export

The resulting cleaned dataset serves as the foundation for subsequent analyses on lifestyle and nutrition indicators across Indian states and districts.


## Loading the Data & Initial inspection

In [1]:
import numpy as np 
import pandas as pd 

df=pd.read_csv("NFHS-5.csv") ## loaded the data set 
df

,Country,State,District,Year,"Women Age Group 15 To 19 Years Who Are Anaemic (%) (UOM:%(Percentage)), Scaling Factor:1","Women Age Group 15 To 49 Years Who Are Anaemic (%) (UOM:%(Percentage)), Scaling Factor:1","Married Women Age Group 15 To 49 Years Use Any Family Planning Methods (%) (UOM:%(Percentage)), Scaling Factor:1","Married Women Age Group 15 To 49 Years Any Modern Family Planning Method (%) (UOM:%(Percentage)), Scaling Factor:1","Average Out Of Pocket Expenditure For Each Delivery In Public Health Facility (UOM:INR(IndianRupees)), Scaling Factor:1","Births Attended By Skilled Health Personnel (%) (UOM:%(Percentage)), Scaling Factor:1",...,"Women Who Are Overweight Or Obese (%) (UOM:%(Percentage)), Scaling Factor:1","Women Who Have High Risk Waist-To-Hip Ratio (%) (UOM:%(Percentage)), Scaling Factor:1","Women With Body Mass Index (Bmi) Below Normal (%) (UOM:%(Percentage)), Scaling Factor:1","Women With 10 Or More Years Of Schooling (%) (UOM:%(Percentage)), Scaling Factor:1","Women Suffering From High Blood Sugar Level (%) (UOM:%(Percentage)), Scaling Factor:1","Women Suffering From High Or Very High Blood Sugar Level (%) (UOM:%(Percentage)), Scaling Factor:1","Women Suffering From Very High Blood Sugar Level (%) (UOM:%(Percentage)), Scaling Factor:1","Women Suffering From Elevated Blood Pressure (%) (UOM:%(Percentage)), Scaling Factor:1","Women-Mildly Elevated Blood Pressure (%) (UOM:%(Percentage)), Scaling Factor:1","Women-Moderately Or Severely Elevated Blood Pressure (%) (UOM:%(Percentage)), Scaling Factor:1"
0,India,Andaman And Nicobar Islands,Nicobars,"Calendar Year (Jan - Dec), 2019",48.0,38.3,65.3,57.2,2278,98.6,...,39.1,62.5,8.2,53.5,7.4,13.1,3.9,35.4,23.2,8.5
1,India,Andaman And Nicobar Islands,North And Middle Andaman,"Calendar Year (Jan - Dec), 2019",47.8,62.1,84.1,73.1,1904,98.3,...,35.9,79.3,8.6,41.0,7.2,16.7,6.4,27.4,18.4,4.0
2,India,Andaman And Nicobar Islands,South Andamans,"Calendar Year (Jan - Dec), 2019",43.2,57.7,57.1,50.5,3460,96.9,...,39.0,78.2,10.0,57.5,7.5,18.4,9.5,23.0,12.7,4.9
3,India,Andhra Pradesh,Ananthapuramu,"Calendar Year (Jan - Dec), 2019",50.0,50.5,67.6,67.6,2682,98.3,...,29.3,46.3,21.9,31.3,5.4,14.2,7.8,23.4,12.4,8.5
4,India,Andhra Pradesh,Chittoor,"Calendar Year (Jan - Dec), 2019",43.9,51.8,69.0,68.9,2533,97.6,...,33.4,41.7,20.1,40.3,6.9,17.2,8.4,22.7,10.7,5.4
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
699,India,West Bengal,Purba Bardhaman,"Calendar Year (Jan - Dec), 2019",78.2,77.9,67.9,58.5,1892,94.3,...,23.9,73.0,16.3,27.3,8.1,16.6,8.1,21.2,11.5,5.7
700,India,West Bengal,Purba Medinipur,"Calendar Year (Jan - Dec), 2019",73.8,70.4,59.3,48.1,4390,94.6,...,27.4,66.2,16.5,30.9,8.8,21.3,11.0,18.1,9.6,4.7
701,India,West Bengal,Purulia,"Calendar Year (Jan - Dec), 2019",79.2,76.7,51.0,41.3,2478,91.4,...,9.6,51.6,33.7,26.9,10.1,15.3,4.8,17.5,9.6,3.9
702,India,West Bengal,South 24 Parganas,"Calendar Year (Jan - Dec), 2019",57.3,61.6,82.9,67.9,2142,93.2,...,21.2,86.0,9.9,34.9,7.2,16.5,8.2,21.5,13.5,5.6


In [2]:
## head and tail 
## df.head()

In [3]:
df.shape 

(704, 108)

 704 ROWS & 108 COLS

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 704 entries, 0 to 703
Columns: 108 entries, Country to Women-Moderately Or Severely Elevated Blood Pressure (%) (UOM:%(Percentage)), Scaling Factor:1
dtypes: float64(101), int64(3), object(4)
memory usage: 594.1+ KB


### Observation

The dataset consists of 704 district-level observations and 108 variables, providing extensive information on health, nutrition, and demographic indicators across India.

## Cloned data set into df_cleann

In [5]:
df_cleann = df.copy()   ## clean data 

In [6]:
df_cleann.columns = (
    df_cleann.columns
    .str.replace(r"\(.*?\)", "", regex=True)
    .str.replace(",", "")
    .str.replace("%", "")
    .str.replace("-", " ")
    .str.replace(r"[^\w\s]", "", regex=True)
    .str.strip()
)

df_cleann.columns = (
    df_cleann.columns
    .str.lower()
    .str.replace(r"\s+", "_", regex=True)
    .str.replace(r"_?scaling_factor\d+", "", regex=True)
)

In [7]:
df_cleann.columns

Index(['country', 'state', 'district', 'year',
       'women_age_group_15_to_19_years_who_are_anaemic',
       'women_age_group_15_to_49_years_who_are_anaemic',
       'married_women_age_group_15_to_49_years_use_any_family_planning_methods',
       'married_women_age_group_15_to_49_years_any_modern_family_planning_method',
       'average_out_of_pocket_expenditure_for_each_delivery_in_public_health_facility',
       'births_attended_by_skilled_health_personnel',
       ...
       'women_who_are_overweight_or_obese',
       'women_who_have_high_risk_waist_to_hip_ratio',
       'women_with_body_mass_index_below_normal',
       'women_with_10_or_more_years_of_schooling',
       'women_suffering_from_high_blood_sugar_level',
       'women_suffering_from_high_or_very_high_blood_sugar_level',
       'women_suffering_from_very_high_blood_sugar_level',
       'women_suffering_from_elevated_blood_pressure',
       'women_mildly_elevated_blood_pressure',
       'women_moderately_or_severely_elev

## Column names were standardized to improve readability and ensure consistency throughout the analysis. Special characters, percentage symbols, brackets, and unnecessary text were removed, while all column names were converted to lowercase and formatted using underscores.

## Duplicate entries check

In [8]:
## duplicates 
df_cleann.duplicated().sum()

0

Zero Duplicates found

In [9]:
#df_cleann.tail()

In [10]:
##df_cleann["year"].value_counts() ## only single year 

## Missing values Analysis
This section examines the extent and distribution of missing values in the dataset. Identifying missing data is an important step in the cleaning process, as it helps determine whether variables should be removed, imputed, or retained for further analysis.

In [11]:
df_cleann.isnull().sum().sort_values(ascending=False)

children_age_group_6_to_8_months_receiving_solid_or_semisolid_food_and_breastmilk                            329
non_breastfeeding_children_age_group_6_to_23_months_receiving_an_adequate_diet                               328
children_with_diarrhoea_in_the_2_weeks_preceding_the_survey_who_received_zinc                                260
children_with_diarrhoea_in_the_2_weeks_preceding_the_survey_who_received_oral_rehydration_salts              260
children_with_diarrhoea_in_the_2_weeks_preceding_the_survey_taken_to_a_health_facility_or_health_provider    260
                                                                                                            ... 
female_population_age_6_years_and_above_who_ever_attended_school                                               0
women_ever_undergone_an_oral_cavity_examination_for_oral_cancer                                                0
women_ever_undergone_a_screening_test_for_cervical_cancer                                       

### Observation

The missing value analysis revealed that most variables contain little to no missing data. However, a small number of child nutrition and healthcare-related indicators exhibited substantial missingness, with the highest missing count reaching 329 observations. These variables were examined separately to determine whether they should be retained, imputed, or removed from the dataset.

In [12]:
missing_percent = df_cleann.isnull().mean() * 100

In [13]:
missing_percent[missing_percent > 0].sort_values()

children_age_group_6_to_23_months_receiving_an_adequate_diet                                                                                         0.142045
deaths_in_the_last_3_years_registered_with_the_civil_authority                                                                                       0.142045
current_users_ever_told_about_side_effects_of_current_method                                                                                         0.142045
children_age_group_9_to_35_months_who_received_vitamin_a_dose_in_the_last_6_months                                                                   0.142045
children_age_5_years_who_attended_pre_primary_school_during_the_year_2019_20                                                                         0.426136
children_age_group_12_to_23_months_who_have_received_3_doses_of_penta_or_hepatitis_b_vaccine                                                         0.710227
children_age_group_12_to_23_months_who_have_received

### Observation

The majority of variables contain very little missing data, with most columns having less than 1% missing values. However, a small number of child nutrition and healthcare-related indicators exhibit substantial missingness. The highest missing percentages were observed in:

- Children aged 6–8 months receiving solid or semi-solid food and breastmilk (46.7%)
- Non-breastfeeding children aged 6–23 months receiving an adequate diet (46.6%)
- Several diarrhoea-related indicators (~36.9%)

Given the high proportion of missing values in these variables, they require special consideration before further analysis.

In [14]:
df_cleann["children_age_group_6_to_8_months_receiving_solid_or_semisolid_food_and_breastmilk"].describe()

count    375.000000
mean       7.051733
std       17.084525
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max       85.200000
Name: children_age_group_6_to_8_months_receiving_solid_or_semisolid_food_and_breastmilk, dtype: float64

In [15]:
col = "children_age_group_6_to_8_months_receiving_solid_or_semisolid_food_and_breastmilk"

nan_count = df_cleann[col].isnull().sum()

zero_count = (df_cleann[col] == 0).sum()

nan_count, zero_count

(329, 311)

In [16]:
df_cleann = df_cleann.drop(
    columns=[
        "children_age_group_6_to_8_months_receiving_solid_or_semisolid_food_and_breastmilk"
    ]
)

### Observation

The variable `children_age_group_6_to_8_months_receiving_solid_or_semisolid_food_and_breastmilk` contained 46.7% missing values (329 observations). In addition, 311 of the remaining observations were recorded as zero, resulting in over 90% of the dataset containing either missing or zero values for this variable. Due to the high level of missingness and limited informational value, the column was removed from the dataset.

In [17]:
#df_cleann.shape ## checking if the column got dropped 

In [18]:
col = "non_breastfeeding_children_age_group_6_to_23_months_receiving_an_adequate_diet"

nan_count = df_cleann[col].isnull().sum()

zero_count = (df_cleann[col] == 0).sum()

non_zero_count = (df_cleann[col] > 0).sum()

nan_count, zero_count, non_zero_count

(328, 317, 59)

In [19]:
df_cleann = df_cleann.drop(
    columns=[
        "non_breastfeeding_children_age_group_6_to_23_months_receiving_an_adequate_diet"
    ]
)

### Observation

The variable `non_breastfeeding_children_age_group_6_to_23_months_receiving_an_adequate_diet` contained 328 missing values (46.6% of observations). Additionally, 317 observations recorded a value of zero, leaving only 59 non-zero observations across the dataset. Since more than 91% of the records contained either missing or zero values, the variable was removed due to its limited analytical value.

In [20]:
## df_cleann.shape

## Exporting the Intermediate Dataset

In [21]:
df_cleann.to_csv("nfhs5_cleaned_partial.csv", index=False)

At this stage, an intermediate version of the cleaned dataset is exported. This checkpoint preserves the progress made during column standardization and the removal of variables with substantial missingness, allowing subsequent cleaning steps to be performed on a separate copy if required.

## Loading the Intermediate data set & Inspection 

In [22]:
import pandas as pd
df_clear = pd.read_csv("nfhs5_cleaned_partial.csv")

In [23]:
df_clear.shape

(704, 106)

In [24]:
df_clear.dtypes

country                                                      object
state                                                        object
district                                                     object
year                                                         object
women_age_group_15_to_19_years_who_are_anaemic              float64
                                                             ...   
women_suffering_from_high_or_very_high_blood_sugar_level    float64
women_suffering_from_very_high_blood_sugar_level            float64
women_suffering_from_elevated_blood_pressure                float64
women_mildly_elevated_blood_pressure                        float64
women_moderately_or_severely_elevated_blood_pressure        float64
Length: 106, dtype: object

In [25]:
missing_percent = (
    df_clear.isnull().sum() / len(df_clear) * 100
).sort_values(ascending=False)

In [26]:
missing_percent[missing_percent > 0]

children_with_diarrhoea_in_the_2_weeks_preceding_the_survey_who_received_zinc                                                                       36.931818
children_with_diarrhoea_in_the_2_weeks_preceding_the_survey_who_received_oral_rehydration_salts                                                     36.931818
children_with_diarrhoea_in_the_2_weeks_preceding_the_survey_taken_to_a_health_facility_or_health_provider                                           36.931818
children_born_at_home_who_were_taken_to_a_health_facility_for_a_check_up_within_24_hours_of_birth                                                   28.693182
children_with_fever_or_symptoms_of_acute_respiratory_infection_in_the_2_weeks_preceding_the_survey_taken_to_a_health_facility_or_health_provider    21.164773
children_under_age_6_months_exclusively_breastfed                                                                                                   13.352273
births_in_private_health_facility_delivered_by_caesa

In [27]:
col = "children_with_diarrhoea_in_the_2_weeks_preceding_the_survey_who_received_zinc"

nan_count = df_clear[col].isnull().sum()
zero_count = (df_clear[col] == 0).sum()
non_zero_count = (df_clear[col] > 0).sum()

nan_count, zero_count, non_zero_count

(260, 230, 214)

In [28]:
col = "children_with_diarrhoea_in_the_2_weeks_preceding_the_survey_who_received_oral_rehydration_salts"

nan_count = df_clear[col].isnull().sum()
zero_count = (df_clear[col] == 0).sum()
non_zero_count = (df_clear[col] > 0).sum()

nan_count, zero_count, non_zero_count

(260, 230, 214)

### Observation

Two diarrhoea-treatment indicators—children receiving zinc supplementation and children receiving oral rehydration salts (ORS)—contained 260 missing values (36.9% of observations). However, both variables also contained 214 non-zero observations, indicating the presence of meaningful information. Unlike previously removed variables, these columns retained sufficient analytical value and were therefore kept in the dataset for further processing.

In [29]:
cols_with_missing = df_clear.columns[df_clear.isnull().any()]
len(cols_with_missing)

25

In [30]:
skew_missing_cols = df_clear[cols_with_missing].skew()

skew_missing_cols.sort_values()

children_age_group_12_to_23_months_who_have_received_bacillus_calmette_guerin                                                                      -7.207027
children_age_group_12_to_23_months_who_received_most_of_their_vaccinations_in_a_public_health_facility                                             -6.640374
children_age_group_12_to_23_months_who_have_received_the_first_dose_of_measles_containing_vaccine                                                  -4.113303
children_age_group_12_to_23_months_who_have_received_3_doses_of_penta_or_diphtheria_tetanus_toxoids_and_pertussis_vaccine                          -3.827303
children_age_group_12_to_23_months_fully_vaccinated_based_on_information_from_vaccination_card_only                                                -3.101908
children_age_group_12_to_23_months_who_have_received_3_doses_of_penta_or_hepatitis_b_vaccine                                                       -2.985343
children_age_group_12_to_23_months_who_have_received_3_dos

### Choosing an Imputation Strategy

To determine an appropriate imputation method, the skewness of variables containing missing values was examined. Several variables exhibited substantial positive or negative skewness, indicating non-symmetric distributions. Therefore, median imputation was selected because it is more robust to skewed data and extreme values than mean imputation.

In [31]:
df_clean = df_clear.fillna(df_clear.median(numeric_only=True))

In [32]:
##df_clean.isnull().sum().sum()

In [33]:
##df_clean.isnull().sum()

In [34]:
df_clean.to_csv("nfhs5_cleaned_final.csv", index=False)

In [35]:
import pandas as pd 
df_clear=pd.read_csv("nfhs5_cleaned_final.csv" )
df_clear.shape


(704, 106)

In [36]:
df_clean.isnull().sum().sum()

0

##  Conclusion

- Column names were standardized to improve consistency and readability.
- No duplicate records were found in the dataset.
- Missing value analysis identified 25 columns with incomplete observations.
- Columns with substantial missingness and limited analytical value were removed.
- Remaining missing values were imputed using the median due to skewed distributions.
- The final cleaned dataset is complete, consistent, and ready for exploratory data analysis.